sklearn/preprocessing/tests/test_encoders.py

```diff
@@ -827,31 +829,37 @@ def test_encoders_has_categorical_tags(Encoder):
     assert 'categorical' in Encoder()._get_tags()['X_types']
 
 
-@pytest.mark.parametrize('input_dtype', ['O', 'U'])
-@pytest.mark.parametrize('category_dtype', ['O', 'U'])
+@pytest.mark.parametrize('input_dtype', ['O', 'U', 'S'])
+@pytest.mark.parametrize('category_dtype', ['O', 'U', 'S'])
 @pytest.mark.parametrize('array_type', ['list', 'array', 'dataframe'])
-def test_encoders_unicode_categories(input_dtype, category_dtype, array_type):
-    """Check that encoding work with string and object dtypes.
+def test_encoders_string_categories(input_dtype, category_dtype, array_type):
+    """Check that encoding work with object, unicode, and byte string dtypes.
     Non-regression test for:
     https://github.com/scikit-learn/scikit-learn/issues/15616
     https://github.com/scikit-learn/scikit-learn/issues/15726
     """
 
+    # Expect incompatible dtypes to raise an exception on fit(X)
+    may_pass = (category_dtype, input_dtype) not in {('S', 'O')}
+    context = raises(ValueError, match="unknown categories", may_pass=may_pass)
```

Comment:
I think it would be clearer to skip here:

```py
    if train_dtype == "O" and category_dtype == "S":
        # unsupported combination
        return
```

and leave the code that `fit`s and `transform`s the same.

Reply:
Currently the test checks that "Found unknown categories" is raised for the unsupported combos. So I just want to confirm that your intent is to remove such checks.

Reply:
Yea, and then have a new test for the unsupported combination. Having a single test handle both error cases and valid cases, makes the test harder to reason able.

Reply:
I suppose the reason I explicitly tested the unsupported combos is that, due to how pytest parametrization is set up, those combos do appear as parametrized tests in the pytest logs. Before this PR I did look for a way to officially skip a particular combo of pytest parametrizations instead, but couldn't find such a feature in pytest.

If it's kosher in the sklearn codebase to have some always-skipped tests, then I agree skipping simplifies the test code.

Reply:
> If it's kosher in the sklearn codebase to have some always-skipped tests, then I agree skipping simplifies the test code.

I agree is a little strange to skip like this. I would prefer to totally remove them from the logs, which requires something like this:

```py
from itertools import product

@pytest.mark.parametrize('input_dtype, category_dtype', 
[(input_dtype, cat_dtype)
 for input_dtype, cat_dtype in product(['O', 'U', 'S'], ['O', 'U', 'S'])
 if (input_dtype,  cat_dtype) != ("O", "S")])
@pytest.mark.parametrize('array_type', ['list', 'array', 'dataframe'])
def test_encoders_unicode_categories(input_dtype, category_dtype, array_type):
```

(This is not the nicest looking code)

Reply:
I can't bring myself to duplicate of this test with only a couple of lines modified from the original. Since the invalid combos were not tested to begin with, would you mind if they're just skipped as you originally proposed?

Reply:
Ah, OK if doing the combos using *product* is acceptable then I'll do that and lose the *if* statement. Thanks for the suggestion.

Reply:
The test this PR is modifying is already a regression test, so we have to be careful with adjusting it. I think adding a separate test just for bytes, should be good enough and would be easier for reviewers.

BTW, I do not think we explicitly support bytes, and previous support was not intentional. When looking at `test_encoders`, we do not have any test for bytes.

Reply:
Right now the test works with exactly two changes to the original based on your suggestions:
```python
# deliberately omit 'OS' as an invalid combo
@pytest.mark.parametrize('input_dtype, category_dtype', ['OO', 'OU',
                                                         'UO', 'UU', 'US',
                                                         'SO', 'SU', 'SS'])
```
and
```python
    if input_dtype == 'S':
        # bytes require th input to be consistent type
        X_test = [[b'a'], [b'a'], [b'b'], [b'a']]
    else:
        X_test = [['a'], ['a'], ['b'], ['a']] 
    X_test = _convert_container(X_test, array_type)
```
*_BaseEncoder* seems to be bytes-aware in at least one test:
https://github.com/scikit-learn/scikit-learn/blob/81102146e35c81d7aab16d448f1c2b66d8a67ed9/sklearn/preprocessing/_encoders.py#L145

Reply:
I would be okay with those changes. I would also add that the test is not a regression test for `https://github.com/scikit-learn/scikit-learn/issues/19677`

Please make the changes so they can be reviewed. I would keep the `np.array` because of my comment here: https://github.com/scikit-learn/scikit-learn/pull/19727#discussion_r599073512